src/core/qgstemporalutils.cpp

```diff
@@ -141,3 +142,58 @@ bool QgsTemporalUtils::exportAnimation( const QgsMapSettings &mapSettings, const
 
   return true;
 }
+
+
+QDateTime QgsTemporalUtils::calculateFrameTime( const QDateTime &start, const long long frame, const double timeStep, const QgsUnitTypes::TemporalUnit timeStepUnit )
+{
+
+  double unused;
+  bool isFractional = fabs( modf( timeStep, &unused ) ) > 0.00001;
```

Comment:
I think we could simplify this approach -- make calculateFrameTime accept a QgsInterval value instead of the separate timeStep/timeStepUnit values, and then retrieve the interval original unit and use that here (falling back on the fractional approach if the original interval unit is unknown unit) . That should also allow us to remove the changes to QgsTemporalNavigationObject API, since we could just keep using the QgsInterval mFrameDuration instead of introducing the new mFrameTimeStep/mFrameTimeStepUnit members.

Reply:
right @nyalldawson, shouldn't we fall back to the fractional approach when the original interval duration is a fraction and not when the original interval unit is unknown unit?

Reply:
@Samweli I think both situations should cause the fallback fractional approach. To clarify:

- If we have a known originalUnit (e.g months), and non fractional original originalDuration: use the new approach
- if we have a known originalUnit (e.g. months) and a fractional originalDuration: use the old approach
- if we have a unknown original unit (i.e. QgsUnitTypes::TemporalUnknownUnit): use the old approach